### PR TITLE
Add mp4 (+ mov, mp3, flac, wav, glb, html) upload support in cm v2 cli

### LINKED ISSUE #1454: Add MP4 uploads to CM
Hey, got skipping unsupported format error when trying to upload mp4 files with CmV2. Checked the issue in some places and found closest way to the solution in this link: 

https://hackmd.io/@archaeopteryx/By4bpbA4F#Upload-MP4s

But it seems it is for v1 and some files mentioned in the link has changed. 

Cannot upload video NFT's, please help.

### LINKED ISSUE #1506: CMv2 uploading .glb files
**Describe the bug**
Candy machine is skipping .glb files while uploading( unsupported file type error) using CM V2

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to '\metaplex\js\packages\cli\src> '
2. Run 'ts-node candy-machine-v2-cli.ts upload -e devnet -k D:\lock\devnetwallet.json -c cmdevnet -cp C:\metaplex\cmsetup.json C:\metaplex\assets'
3. See an error

**Expected behaviour**
The upload should be fine because in Metaplex docs it's stated that CM supports .glb files

**Screenshots**
image

**Desktop :**
 - OS: [Windows 10]
 - Version [Enterprise 64-bit build 19041]

**Additional context**
Trying test upload on devnet with 2 files 
image
Metadata example 0.json
here are the community tweaks i tried
image
image

### LINKED ISSUE #1545: Add MP4 support to CMV2 CLI
#1506

### PR DESCRIPTION
**Pull request by guillemap**
Fixes #1545 
Fixes #1454 
Fixes #1506

<br />

**/assets** must contain 3 files for each asset: image + animation + metadata. Supported formats:
- Image: PNG, GIF and JPG
- Animation: MP4, MOV, MP3, FLAC, WAV, GLB, HTML
- Metadata: JSON

<br />

**.json** files must add:
`animation_url` - path to a multi-media attachment of the asset. The supported file formats are MP4 and MOV for video, MP3, FLAC and WAV for audio, GLB for AR/3D assets, and HTML for HTML pages.
`properties.category` - Supported categories:
- "image" - PNG, GIF, JPG
- "video" - MP4, MOV
- "audio" - MP3, FLAC, WAV
- "vr" - GLB (3D Models)
- "html" - HTML pages; scripts and relative paths within the HTML page are also supported
 
`properties.files` - Object array, where an object should contain the uri and type of the file that is part of the asset. The type should match the file extension. Supported file types are:
```
  'image/png'
  'image/gif'
  'image/jpeg'
  'video/mp4'
  'video/quicktime'
  'audio/mpeg'
  'audio/x-flac'
  'audio/wav'
  'model/gltf-binary'
  'text/html'
```
 The array should also include files specified in `image` and `animation_url` fields.

<br />

0.json **example** structure for a video animation:

```
{
  "name": "",
  "symbol": "",
  "description": "Test NFT",
  "seller_fee_basis_points": 500,
  "image": "0.png",
  "animation_url": "0.mp4",
  "attributes": [
    {
      "trait_type": "background",
      "value": "blue"
    }
  ],
  "collection": {
    "name": "",
    "family": ""
  },
  "properties": {
    "files": [
      {
        "uri": "0.png",
        "type": "image/png"
      },
      {
        "uri": "0.mp4",
        "type": "video/mp4"
      }
    ],
    "category": "video",
    "creators": [
      {
        "address": "",
        "share": 100
      }
    ]
  }
}
```

### CODE REVIEW

**Review by irs-team — APPROVED** [review]
Thank you for this! Tested with IPFS & Arweave

**Comment by stegaBOB** [comment]
Awesome work! Going to run some tests over the next couple of days and then hopefully merge it in early next week!

**Comment by guillemap** [comment]
Awesome! Thanks 🎉

**Comment by wlemahieu** [comment]
@guillemap This looks great! Thanks for putting in the time to do this.

I tested this PR using a fresh copy of the latest Metaplex. There appears to be a bug with "arweave-bundle" storage and how it's writing .HTML and .PNG URIs to the manifest after uploading. The problem is the .HTML file is put in the "image" property's place, and the .PNG file URI is not set in the manifest at all. The animation_url is left as "0.html" as well.

Here is a basic example:

https://arweave.net/Wh4ZnkHChM2blU-4wF9ppISoBpXy5sW0-mF18sUv--c

This issue is not happening when using "arweave" storage option and both the PNG and HTML are properly. However, Metaplex remote server seems to compress HTML files into DMS files effectively breaking them, but since that storage option is going away it shouldn't matter. But this points to the issue being somewhere in the ARBundle portion of the Metaplex code.

I'll take a look to see if I can see where the bug is occurring, but just figured I'd mention my PR test results in relation to .HTML animation_url here for @stegaBOB.

This should be easily reproduced with a single asset inside /assets (0.png, 0.html, 0.json) and using the above .json structure. Of course I tested this on mainnet-beta since real AR is necessary. 

Let me know if I can provide any other info. If I find the source of the bug I will update this thread.

**Comment by guillemap** [comment]
Thanks for pointing this out @wlemahieu!

The way I approached the solution was to iterate twice through image & animation uploads (both uploading a media file & a json) right on the upload module. This way storage modules code were not changed. However Bundle upload was ignored, and sure it's not ideal.

I'm refactoring the approach by coding the animation upload on each storage module, so we get two improvements:
- animation upload (including .html) works for all storage modes including `arweave-bundle`
- only one json is uploaded, with all links properly set. Reducing upload costs and increasing overall speed

**Review by ethanitovitch — COMMENTED** [review]
Looks good to me :)

File: js/packages/cli/src/helpers/upload/ipfs.ts [hunk 1]
```diff
@@ -40,10 +41,35 @@ export async function ipfsUpload(
 
   await sleep(500);
 
+  let animationUrl = undefined;
```
Comment (roederw) [inline js/packages/cli/src/helpers/upload/ipfs.ts:44]: Most of this is copy paste from the `imageHash` work above. Can you refactor this?

Comment (guillemap) [reply js/packages/cli/src/helpers/upload/ipfs.ts]: Sure, refactored aws as well

**Review by tusharLamba11 — APPROVED** [review]
Great work. I tried to deploy mp4 NFTs and its perfectly working.

**Comment by igorbarbashin** [comment]
@guillemap I believe it breaks the initial upload when there's no cache file at all. I just ran into this issue and investigating

**Review by stegaBOB — APPROVED** [review]
Removed the ability to upload animations via "arweave". Other than that, everything was awesome!

**Pull request merged**